eng/bomgenerator/src/main/java/com/azure/tools/bomgenerator/DependencyAnalyzer.java

```diff
@@ -0,0 +1,238 @@
+        if (!dependencyTree.containsKey(dependency)) {
+            dependencyTree.put(new BomDependencyNoVersion(dependency.getGroupId(), dependency.getArtifactId()), new HashMap<>());
+        }
+
+        HashMap<String, Collection<BomDependency>> versionToParents = dependencyTree.get(dependency);
+        if(!versionToParents.containsKey(dependency.getVersion())) {
+            versionToParents.put(dependency.getVersion(), new ArrayList<>());
+        }
+
+        if(parentDependency != null) {
+            versionToParents.get(dependency.getVersion()).add(parentDependency);
+        }
+    }
+
+
+    private void makeDependencyInEligible(BomDependency dependency, String dropDependencyReason) {
+        if(nameToVersionToChildrenDependencyTree.containsKey(dependency)) {
+            HashMap<String, Collection<BomDependency>> versionToDependency = nameToVersionToChildrenDependencyTree.get(dependency);
+            bomIneligibleDependencies.put(dependency, dropDependencyReason);
+            if(dropDependencyReason == null) {
+                dropDependencyReason = dependency.toString();
+                logger.trace("\t\tDropping dependency {}", dependency.toString());
+            }
+            else {
+                logger.trace("\t\tDropping dependency {} due to dependency {}", dependency.toString(), dropDependencyReason);
+            }
+
+            // Make all the dependencies that include these also ineligible.
+            String finalDropDependencyReason = dropDependencyReason;
+            versionToDependency.get(dependency.getVersion()).forEach(parent -> makeDependencyInEligible(parent, finalDropDependencyReason));
+        }
+    }
+
+    private void pickCoreDependencyRoots() {
+        BomDependency coreDependency = inputDependencies.stream().filter(dependency -> dependency.getArtifactId().equals("azure-core")).findFirst().get();
+        coreDependencies.put(BomDependency.convertTo(coreDependency), coreDependency);
+        coreDependencies.putAll(getDependencies(coreDependency).stream().collect(Collectors.toMap(dependency -> BomDependency.convertTo(dependency), dependency -> dependency)));
+        this.bomEligibleDependencies.addAll(coreDependencies.values().stream().filter(dependency -> !externalDependencies.contains(dependency)).collect(Collectors.toList()));
+    }
+
+    private void resolveConflict(BomDependencyNoVersion dependencyNoVersion) {
+        Map<String, Collection<BomDependency>> versionToDependency = nameToVersionToChildrenDependencyTree.get(dependencyNoVersion);
+        if (versionToDependency.size() > 1) {
+            List<String> versionList = versionToDependency.keySet().stream().sorted(new DependencyVersionComparator()).collect(Collectors.toList());
+            String eligibleVersion;
+
+            logger.trace("Multiple version of the dependency {} included", dependencyNoVersion);
```

Comment:
I'm a bit confused here, is the above map an inverted map which contains all the versions of a given dependency, basically the structure:

```
<dependency name>
  <version>
    <libraries that use version>
  <version 2>
    <libraries that use version 2>
...
```

Reply:
Yes - the resolveTree actually creates the map of this nature. Basically an n-ary tree with all the ancestors pointers.